url8af84d296f_mrcosta_gantt_generator_tgz-pJ8-org_gantt_generator_mrcpsp_RunJ8/src/org/swiftgantt/common/SwingComImageGenerator.java

```diff
@@ -0,0 +1,190 @@
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+
+import javax.imageio.spi.ImageWriterSpi;
+import javax.imageio.stream.FileImageOutputStream;
+import javax.imageio.stream.ImageOutputStream;
+import javax.swing.JComponent;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import com.sun.image.codec.jpeg.JPEGCodec;
+import com.sun.image.codec.jpeg.JPEGImageEncoder;
+import com.sun.imageio.plugins.png.PNGImageWriter;
+import com.sun.imageio.plugins.png.PNGImageWriterSpi;
+
+/**
+ * Swing Component Image Generator.
+ * @author Yuxing Wang
+ *
+ */
+public class SwingComImageGenerator {
+	private static SwingComImageGenerator instance = new SwingComImageGenerator();
+	private Logger logger;
+	
+	private SwingComImageGenerator() {
+		logger = LogManager.getLogger(this.getClass());
+	}
+	
+	/**
+	 * 
+	 * @return
+	 */
+	public static SwingComImageGenerator getInstance(){
+		return instance;
+	}
+	/**
+	 * 
+	 * @param component
+	 * @param filePath
+	 * @throws Exception
+	 */
+	public void genJPEGImage(JComponent component, String filePath) throws IOException{
+		Rectangle bounds = component.getBounds();
+		this.genJPEGImage(new Dimension(bounds.width, bounds.height), new JComponent[]{component}, new Rectangle[]{bounds}, filePath);
+	}
+	
+	/**
+	 * 
+	 * @param size
+	 * @param components
+	 * @param areas
+	 * @param filePath
+	 * @throws IOException
+	 */
+	public void genJPEGImage(Dimension size, JComponent[] components, Rectangle[] areas, String filePath) throws IOException{
+		logger.debug("Generate JPEG image file");
+		if(components.length == 0 || components.length != areas.length){
+			throw new RuntimeException("The specified components is not fulfil the requirement");
+		}
+		if(filePath == null){
+			String dir = System.getProperty("user.dir") + System.getProperty("file.separator") + "image";
+			filePath = dir + System.getProperty("file.separator") + "ganttchart.jpg";
+		}
+		checkAndCreateDirForFilePath(filePath);
+		BufferedImage image = new BufferedImage(size.width, size.height, BufferedImage.TYPE_INT_RGB);
+		Graphics imageGraphic = image.getGraphics();
+		// Background is white.
+		imageGraphic.setColor(Color.WHITE);
+		imageGraphic.fillRect(0, 0, size.width, size.height);
+
+		for (int i = 0; i < components.length; i++) {
+			JComponent curCom = components[i];
+			Rectangle bounds = areas[i];
+			if (logger.isDebugEnabled()) {
+				logger.debug("generateJPEGImage(Dimension, JComponent[], Rectangle[], String) - Start to paint component " + curCom.getClass().getName() + " to image:" + bounds); //$NON-NLS-1$ //$NON-NLS-2$
+			}
+			curCom.paint(imageGraphic);
+			if(bounds.x > 0 || bounds.y > 0){
+				// Copy may be not effective.
+				imageGraphic.copyArea(0, 0, bounds.width, bounds.height, bounds.x, bounds.y);
+			}
+		}
+		try{
+			File imgFile = new File(filePath);
+			FileOutputStream fos = new FileOutputStream(imgFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at lines 96, 97. The resource is closed at line 99. There are other execution paths that do not contain closure statements, for example, when JPEGCodec.createJPEGEncoder throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).